assessment/collectors/project_analyzer/file_dependency_analyzer.go

```diff
@@ -211,25 +212,108 @@ func (b *BaseAnalyzer) TopologicalSort(G map[string]map[string]struct{}) ([][]st
 	for node := range G {
 		inDegree[node] = 0
 	}
-	var maxDegree int
 
 	for node := range G {
 		for neighbor := range G[node] {
 			inDegree[neighbor]++
-			if inDegree[neighbor] > maxDegree {
-				maxDegree = inDegree[neighbor]
-			}
 		}
 	}
 
-	taskLevels := make([][]string, maxDegree+1)
-
+	// Use Kahn's algorithm
+	queue := []string{}
 	for node, degree := range inDegree {
-		degree = maxDegree - degree
-		taskLevels[degree] = append(taskLevels[degree], node)
+		if degree == 0 {
+			queue = append(queue, node)
+		}
+	}
+
+	sortedTasks := []string{} // Changed to a simple string slice
+	for len(queue) > 0 {
+		// Dequeue a node
+		node := queue[0]
+		queue = queue[1:]
+		sortedTasks = append([]string{node}, sortedTasks...)
+
+		// For each neighbor of the dequeued node
+		for neighbor := range G[node] {
+			// Decrease the in-degree of the neighbor
+			inDegree[neighbor]--
+			// If the in-degree of the neighbor becomes 0, enqueue it
+			if inDegree[neighbor] == 0 {
+				queue = append(queue, neighbor)
+			}
+		}
+	}
+
+	// Check for cycles. If the result doesn't contain all nodes, there's a cycle.
+	if len(sortedTasks) != len(G) {
+		return nil, fmt.Errorf("graph contains a cycle")
 	}
 
-	return taskLevels, nil
+	groupedTasks := groupTasksOptimized(sortedTasks, G)
+	return groupedTasks, nil
+}
+
+// groupTasksOptimized groups tasks based on their dependencies in a directed acyclic graph (DAG).
+//
+// This function efficiently groups tasks into independent sets, ensuring that
+// tasks within the same group do not have dependencies on each other. This is
+// useful for determining parallel execution opportunities and visualizing task
+// dependencies.
+//
+// Args:
+//
+//	tasks: A list of tasks in topological order. Tasks should be listed
+//	       before their dependencies.
+//	graph: A map representing the task dependencies. An edge (u, v)
+//	       indicates that task u must be completed before task v.
+//
+// Returns:
+//
+//	A list of lists, where each inner list represents a group of independent
+//	tasks.
+//
+// Complexity:
+//
+//	Time Complexity: O(n * m), where n is the number of tasks and m is the
+//	    average number of dependencies per task.
+//	Space Complexity: O(n) to store the task-to-group mapping.
+//
+// Example:
+//
+//	// Example usage (assuming you have a graph representation in Go)
+//	// graph := map[string]map[string]struct{}{
+//	//     "1": {"2": {}, "3": {}},
+//	//     "2": {"4": {}},
+//	//     "3": {"4": {}},
+//	//     "4": {},
+//	// }
+//	// sortedTasks := []string{"1", "2", "3", "4"}
+//	// groupedTasks := groupTasksOptimized(sortedTasks, graph)
+//	// // groupedTasks will be: [][]string{{"1"}, {"2", "3"}, {"4"}}
+func groupTasksOptimized(tasks []string, graph map[string]map[string]struct{}) [][]string {
```

Comment:
This is already done at the place where table dependencies via foreign keys and interleaving are resolved. Can we use the same logic instead of re-writing it ?

Reply:
Here we are actually not resolving dependencies but grouping dependencies in such a way that they can be processed concurrently.  The code in Utils it a bit different.